design-proposal.md

```diff
@@ -0,0 +1,66 @@
+# Proposal: Design of the rkt + Kubernetes CRI
+
+## Abstract
+
+This proposal can be viewed as more a discussion than an actual proposal, because we still haven't reached a consensus on the architecture.
+
+## Backgroud
+
+As container runtime interface (CRI) proposals and original stub codes are already merged into Kubernetes, the actual work has already begun.
+All existing and future container runtimes are expected to conform the interface and provides feature parities.
+
+## Design(s)
+
+There are three different designs:
+
+- `Kubelet` --_gPRC calls_--> `rktlet service` --_invoke 'rkt' binary, or use some rkt library_--> result
+
+- `Kubelet` --_in process invoke_--> `rktshim` --_gPRC calls_--> `rkt API service` --> result
+
+- `Kubelet` --_gPRC calls_--> `rktlet service` --_gPRC calls_--> `rkt API service` --> result
+
+1. The major difference of the first two approaches is the 1st approach doesn't require any rkt code to live in the kubernetes codebase, which improves the Kubelet's modularity as a whole.
+However, it appears the rktlet code will live outside of rkt project as well since the gRPC interfaces are Kubernetes specific.
+This means for the 1st approach, we protentially have more work to do, and many of them are duplicated effort with the current rkt api service.
+Also I don't think fork/exec a rkt process is the best ideal state for the rktnetes integration, so I would expect we export a lot of rkt functionalities as a library.
+
+2. For the second approach, it will require some shim code compiled with Kubelet so as to convert Kubernetes pod/container operations into rkt pod/app operations.
+Other than that, we can have the actual CRI implementation in the rkt API service, then we get the setup code and the rkt internal data structrue, functions for free.
+
+3. The third approach is somehow a combination of the first two.
+It has the benefits of the both worlds, rktlet lives outside of Kubelet, so it could lead to higher iteration speed, smaller Kubelet size, and better modularity.
+Also on the other side, the 'rkt API service' is compiled with rkt, and communicate with the rktlet through gRPC, so we still can use a lot of the existing rkt facilities.
+The potential con is that instead of one gPRC service, now we introduce two, which may or may not cause some overhead.
```

Comment:
Y, there is a [measurement](https://docs.google.com/spreadsheets/d/1w1OQ0eoRJvIt4QxsZyKV7-h-hXldpLgDU8_wKZgKm0g/edit#gid=0) of logging api on json-based api, but gRPC should have better performance.